src/Fixer/FunctionNotation/NoSpacesAfterFunctionNameFixer.php

```diff
@@ -145,40 +133,15 @@ private function getFunctionyTokenKinds()
     {
         static $tokens = [
             T_ARRAY,
-            T_ECHO,
             T_EMPTY,
             T_EVAL,
             T_EXIT,
-            T_INCLUDE,
-            T_INCLUDE_ONCE,
             T_ISSET,
             T_LIST,
-            T_PRINT,
-            T_REQUIRE,
-            T_REQUIRE_ONCE,
             T_UNSET,
             T_VARIABLE,
```

Comment:
should this list contain `T_DECLARE` as well, like `declare     (ticks=1);` ?

Reply:
Actually I would do the opposite and handle it in #4435 instead.

Reply:
- I have always seen `declare(strict_types=1);`, never `declare (strict_types=1);`. Also, unlike `echo`/`include`, for `declare` parentheses *are* meaningful (and required). So it looks like it could be added to this list...
- But the PHP manual documents it under "[Control Structures](https://www.php.net/manual/en/control-structures.declare.php)", alongside `if` and `switch` (for which parentheses are required too), and [PSR-2 says](https://www.php-fig.org/psr/psr-2/#5-control-structures) "There MUST be one space after the control structure keyword"...
- However, [PSR-12 says](https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements) "Declare statements MUST contain no spaces".

So, what about handling that as a separate issue?

Reply:
I don't like all these exceptions, why `T_LIST` and not `T_DECLARE`, all arguments made here can done the other way around by someone else next week and than we are back here.

`T_LIST is a language construct not a function, why is touched, PSR says nothing about language constructs` 

So I'm -1 for any language construct here and +1 for a new fixer for these with opt-in for each so everyone can be happy.

Reply:
@SpacePossum: Fair point (the problem actually dates back to the initial creation of FunctionCallSpaceFixer in #683), and good idea 🙂 Just 2 questions:
- You would keep `T_VARIABLE` in this list (i.e. still fix `$f (1)` to `$f(1)`), wouldn't you?
- Do you want me to update this PR accordingly, or would it need a new PR?

Thanks

Reply:
As there is no consensus on other keywords handling, let's postpone it for now. We can merge this PR while we discuss what we'll do next.

Reply:
ok, please restore the include and require than, because `require_once(__ROOT__.'/config.php');` looks valid to me and it is also was not part of the issue

Reply:
@SpacePossum: Actually `include`/`require`[`_once`] *are* part of the issue like `echo`/`print`
<sub>(maybe even "more" since they are documented as "Control Structures", not just "language constructs")</sub>.

You can write
`require_once(__DIR__.'/config.php');` if you prefer (by the way, it will be fixed to
`require_once __DIR__.'/config.php';` with `@Symfony` or `@PhpCsFixer` ruleset, due to the additional `include` rule compared to `@PSR2`).

But "fixing" the valid
`$config = require_once ($dir ?: __DIR__).'/config.php';` to
`$config = require_once($dir ?: __DIR__).'/config.php';` is *objectively wrong*. (I insist, that's not a matter of preference. One could prefer that `include` & friends work like usual functions regarding parentheses around args, but the objective fact is that they don't.)

For example (and sorry if I seem to be repeating myself), given a `function my_incl($file) { return include $file; }`, then
`$r = my_incl($a ?: $b).$c;` is evaluated as `$r = (my_incl($a ?: $b)).$c;`, but
`$r = include($a ?: $b).$c;` is evaluated as `$r = include(($a ?: $b).$c);` (and either could be right or wrong depending on the values of `$a`, `$b` and `$c`, e.g. https://3v4l.org/IrrSl or https://3v4l.org/WIFtY).

https://www.php.net/manual/en/function.include.php warns about it (search "parentheses" and see **Example #<i></i>4**). A CS fixer should not change a valid writing to a misleading one.

I have completed the issue title as per the discussion, and the PR tests, to make it clear.

I hope you understand, thanks

Reply:
I think there is a lot of confusion here so I try to summarize:
- I don't think there is a bug
- you and Julien like to change print and echo and I agreed
- the PR does not only do that, it does more

as such:
- I vote against the PR
- the PR can go in without my vote if you find another, or if you change the PR to the agreed specs

Reply:
and to be even more clear, I _understand_ what you are saying all the time, just because I don't agree doesn't mean I don't understand it

Reply:
> I think there is a lot of confusion here so I try to summarize:

I appreciate that. Outside of different opinions, I find your answers have often been enlightening.

>   * I don't think there is a bug

Well, whichever "bugfix" (for a "bug report") or "enhancement" (for a "feature request"), I would actually be fine as long as it gets merged :)

>   * you and Julien like to change print and echo and I agreed
>   * the PR does not only do that, it does more

The issue discussion progressed gradually; to recap:

1. I initially reported an unwanted change I encountered on a line of the form
   "echo (A op1 B) op2 C;" <sup>(where op2 has a higher precedence than op1 and is not commutative, hence the parentheses are needed and the operands cannot be flipped to "echo C op2 (A op1 B);")</sup>.
   The description didn't try to do further analysis (maybe it should have?).
2. Your first answer made me think to try with `print`, which exhibited the same behavior (and I added "print" besides "echo" in the title).
3. Your second answer made me then realize that the issue was also affecting `include` and `require` (but I didn't update the title...).
4. When I got a bit more time, I explored the code and found that there is already a special case precisely for echo/print and include/require[_once], but only if op2 is `?` (not `.`, `/` etc.), which reinforced my conviction.
5. Finally you proposed that someone make a PR, which I did on the next weekend ;) and there we are.

> as such:
> 
>   * I vote against the PR
>   * the PR can go in without my vote if you find another, or if you change the PR to the agreed specs

I find that a bit sad, and even if @julienfalque has already approved the PR, I would prefer an outcome that satisfies everyone...

---

> and to be even more clear, I _understand_ what you are saying all the time, just because I don't agree doesn't mean I don't understand it

I had no intention of suggesting that you don't understand what I'm saying, please accept my apologies! I meant to express the hope that you would accept that I don't "restore the include and require" (and why).

---

Finally, I still agree with your:

> So I'm -1 for any language construct here and +1 for a new fixer for these with opt-in for each so everyone can be happy.

but I also agree with Julien that it would be preferable as a subsequent issue/PR.

---

Sorry for the length 😟 Did I forget something (again)?

Reply:
I consider these changes going from one preference to another, as such it is a BC break. Also anyone can reason the _after_ this PR to do additional changes or revert the changes. This is not maintainable.

We can avoid it by making the language constructs configurable in the next minor release. This has my preference and that is all there is to it. Maybe @kubawerlos  and @keradus can state what they think?

Reply:
To be fair, the "BC break" here is only that php-cs-fixer would no longer do a controversial change (and I would argue that if it is controversial, then it should not do it, at least not by default)...

But I agree with you (by the way, I just targeted 2.15 as the lowest applicable branch, knowing it would be easier to re-target it to an higher branch), and am looking forward to suggestions from @kubawerlos and @keradus 👍

Reply:
What about this fixer the handle only functions? I mean to remove `T_EVAL` and others as well, and https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4435 could be renamed to `SpaceAfterConstructFixer`, with the config similar to `binary_operator_spaces` - `default` plus exceptions, so PSR12 config would look like:
```php
    'default' => 'single',
    'constructs' => [
        'declare' => 'none',
        'eval' => 'none',
    ]
```

Reply:
`eval()` is a language construct but it acts exactly like a function from user point of view, so I would keep current behavior.

Reply:
> from user point of view

I believe that's exactly the problem we discuss here - "the point of view" - for some `require` can be a function (as it can be called with parenthesis) and for others not. Unless we take this from PHP tokens point of view there always be doubts about what is the function and what is not.

Reply:
That's why it should be covered by a dedicated rule or a new configuration option so out of scope of this PR.

Reply:
Thanks, so we all think `require` is out of scope of this PR, so the change must be reverted

Reply:
See #4859.

Reply:
._. Sorry but what's the argument for reverting require/include but not print?
(By the way, if I had instead opened an issue for require, then added include [and their _once variants], and then realized echo/print have the same problem [both conceptually and in the actual fixer code], would you be demanding to revert echo/print [but not include]?)

Even if my actually encountered case (an echo) would still be solved, to me it doesn't make sense to handle one but not the other... Tomorrow someone may open an issue for include/require... (actually I probably will [and split this PR], if you really want to limit this PR to echo/print)

Anyway I'm also looking forward to outcomes (which may supersede this PR) in #4859

Reply:
I do think that `require` (and similar keywords) should be treated like `echo` but not everyone agrees. That includes @SpacePossum but we can safely assume that more people not involved in the discussion here will disagree too and consider this change a regression.

Actually this reasoning might even apply to `echo`, maybe better postponing this PR until we agree to a solution in #4859?

Reply:
>Sorry but what's the argument for reverting require/include but not print?

I'm sorry, I try again:
- if someone want to change something here, the author must make the argument for the change, not the reviewer
- up to now the argument that has been made over and over again in all kinds of different wording was; I don't like what the fixer does, I have a different preference. I say preference because I haven't seen any evidence the fixer does not what it should do based on the agreed workings when it was created
- if you want to change how it works, we keep the BC promise and do it on the next minor

I'm gonna say, I don't like even to see `print` and `echo` changed anymore. Simplify because I'm done defending the current fixer.

Reply:
I agree with your points, except for this part:

> I haven't seen any evidence the fixer does not what it should do based on the agreed workings when it was created

I find https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/683#discussion_r19931222 and https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1237#issuecomment-112364867 (both ignored, regrettably) are evidence of disagreements as soon as it was created...

Anyway, since you closed this PR, I hope to see constructive propositions in #4859

Reply:
Why are you ignoring me telling you that you can have what you want just on the next minor that will be released at the same time as the next patch release? Can you please respond to the people who try to give you what you want?

Reply:
No offense, I said I agree with your [other] points, including "BC promise" and "next minor". I'm simply waiting for #4859 to define exactly what to do (for the next minor, hopefully) 🙂